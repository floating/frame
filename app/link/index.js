import uuid from 'uuid/v4'
import EventEmitter from 'events'

const source = 'tray:link'

const unwrap = v => v !== undefined || v !== null ? JSON.parse(v) : v
const wrap = v => v !== undefined || v !== null ? JSON.stringify(v) : v

const handlers = {}

const link = new EventEmitter()
link.rpc = (...args) => {
  let cb = args.pop()
  if (typeof cb !== 'function') throw new Error('link.rpc requires a callback')
  let id = uuid()
  handlers[id] = cb
  window.postMessage(wrap({ id, args, source, method: 'rpc' }), '*')
}
link.send = (...args) => window.postMessage(wrap({ args, source, method: 'event' }), '*')

window.addEventListener('message', e => {
  let data = unwrap(e.data)
  let args = data.args || []
  if (e.origin === 'file://' && data.source !== source) {
    if (data.method === 'rpc') {
      if (!handlers[data.id]) return console.log('link.rpc response had no handler')
      handlers[data.id](...args)
      delete handlers[data.id]
    } else if (data.method === 'event') {
      if (!data.channel) return console.log('link.on event had no channel')
      link.emit(data.channel, ...args)
    }
  }
}, false)

export default link
