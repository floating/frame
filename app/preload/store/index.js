/* globals fetch */

import { ipcRenderer } from 'electron'

import EventEmitter from 'events'
import Restore from 'react-restore'

import * as actions from './actions'
import state from './state'

import rpc from '../rpc'

import PersistStore from 'electron-store'

const persist = new PersistStore()

export const store = Restore.create(state(), actions)
store.events = new EventEmitter()

rpc('getSigners', (err, signers) => {
  if (err) return store.signersError(err)
  store.updateSigners(signers)
})

rpc('launchStatus', (err, status) => {
  if (err) return console.log(err) // launchStatusError
  store.setLaunch(status)
})

ipcRenderer.on('main:addSigner', (e, signer) => store.addSigner(signer))
ipcRenderer.on('main:removeSigner', (e, signer) => {
  if (store('signer.current') === signer.id) store.unsetSigner()
  store.removeSigner(signer)
})
ipcRenderer.on('main:updateSigner', (e, signer) => store.updateSigner(signer))
ipcRenderer.on('main:setSigner', (e, signer) => {
  if (signer.id) {
    store.setSigner(signer)
  } else {
    store.unsetSigner()
  }
})

const etherRates = () => {
  fetch('https://api.coinbase.com/v2/exchange-rates?currency=ETH').then(res => res.json()).then(res => {
    if (res && res.data && res.data.rates) store.updateExternalRates(res.data.rates)
  }).catch(e => console.log('Unable to fetch exchange rate', e))
}
etherRates()
setInterval(etherRates, 10000)

// Store Observers
let network = ''
store.observer(() => {
  if (network !== store('local.connection.network')) {
    network = store('local.connection.network')
    ipcRenderer.send('tray:setNetwork', network)
  }
})

store.observer(_ => persist.set('local', store('local')))

let launch = store('local.launch')
store.observer(() => {
  if (launch !== store('local.launch')) {
    launch = store('local.launch')
    if (launch) {
      rpc('launchEnable', err => console.log(err))
    } else {
      rpc('launchDisable', err => console.log(err))
    }
  }
})

export default store
