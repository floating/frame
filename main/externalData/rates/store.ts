import log from 'electron-log'
import { AssetType } from '@framelabs/pylon-client/dist/assetId'

import { storeApi } from '../storeApi'
import { toTokenId } from '../../../resources/domain/balance'

import type { AssetId } from '@framelabs/pylon-client/dist/assetId'
import type { Rate, WithTokenId } from '../../store/state/types'

type RateUpdate = {
  id: AssetId
  data: {
    usd: number
    usd_24h_change: number
  }
}

const RATES_EXPIRY = 1000 * 5 * 60 // rates are valid for 5 minutes before being considered stale
const timeouts: Record<string, NodeJS.Timer> = {}

const separateUpdates = (updates: RateUpdate[]) =>
  updates.reduce(
    (acc, update) => {
      return update.id.type === AssetType.NativeCurrency
        ? { ...acc, native: [...acc.native, update] }
        : { ...acc, token: [...acc.token, update] }
    },
    { native: [] as RateUpdate[], token: [] as RateUpdate[] }
  )

const gatherTokenRates = (updates: RateUpdate[]) =>
  updates.reduce((rates, update) => {
    const tokenId = toTokenId(update.id as WithTokenId)

    rates[tokenId] = {
      usd: {
        price: update.data.usd,
        change24hr: update.data.usd_24h_change
      }
    }

    return rates
  }, {} as Record<string, Record<string, Rate>>)

const handleNativeCurrencyUpdates = (updates: RateUpdate[]) => {
  log.debug('Handling native currency rate updates', { chains: updates.map((u) => u.id.chainId) })

  updates.forEach((u) => {
    storeApi.setNativeCurrencyRate(u.id.chainId, {
      price: u.data.usd,
      change24hr: u.data.usd_24h_change
    })

    const id = `native:${u.id.chainId}`
    const expiry = setTimeout(() => storeApi.removeNativeCurrencyRate(u.id.chainId), RATES_EXPIRY)

    clearTimeout(timeouts[id])
    timeouts[id] = expiry
  })
}

const handleTokenUpdates = (updates: RateUpdate[]) => {
  log.debug('Handling token rate updates', { addresses: updates.map((u) => u.id.address) })

  const rates = gatherTokenRates(updates)
  storeApi.setTokenRates(rates)

  updates.forEach((u) => {
    const address = u.id.address as string
    const expiry = setTimeout(() => storeApi.removeTokenRate(address), RATES_EXPIRY)

    clearTimeout(timeouts[address])
    timeouts[address] = expiry
  })
}

const handleUpdates = (rates: RateUpdate[]) => {
  const { token, native } = separateUpdates(rates)

  if (native.length) {
    handleNativeCurrencyUpdates(native)
  }

  if (token.length) {
    handleTokenUpdates(token)
  }
}

export { handleUpdates }
