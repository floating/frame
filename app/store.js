/* globals */

import EventEmitter from 'events'
import Restore from 'react-restore'

import link from '../resources/link'

// const actions = {
//   initialSignerPos: (u, pos) => u('selected.position.initial', () => pos),
//   initialScrollPos: (u, pos) => u('selected.position.scrollTop', () => pos)
// }

import * as actions from '../resources/store/actions.panel'

export default (state, _cb) => {
  const store = Restore.create(state, actions)
  store.events = new EventEmitter()

  // Feed for relaying state updates
  // store.api.feed((state, actions, obscount) => {
  //   actions.forEach(action => {
  //     action.updates.forEach(update => {
  //       // console.log(update)
  //       // if (update.path.startsWith('main')) return
  //       // if (update.path.startsWith('panel')) return
  //       // link.send('tray:syncPath', update.path, update.value)
  //     })
  //   })
  // })

  link.on('action', (action, ...args) => {
    if (store[action]) store[action](...args)
  })

  return store
}
