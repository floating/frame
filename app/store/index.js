/* globals */

import EventEmitter from 'events'
import Restore from 'react-restore'

import link from '../../resources/link'
import * as actions from './actions'

export default (state, cb) => {
  const store = Restore.create(state, actions)
  store.events = new EventEmitter()

  // Feed for relaying state updates
  // store.api.feed((state, actions, obscount) => {
  //   actions.forEach(action => {
  //     action.updates.forEach(update => {
  //       // console.log(update)
  //       // if (update.path.startsWith('main')) return
  //       // if (update.path.startsWith('panel')) return
  //       // link.send('tray:syncPath', update.path, update.value)
  //     })
  //   })
  // })

  link.on('action', (action, ...args) => { if (store[action]) store[action](...args) })
  link.send('tray:ready') // turn on api

  link.send('tray:refreshMain')

  return store
}
