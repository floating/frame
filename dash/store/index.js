// import Restore from 'react-restore'
// import * as actions from './actions'
// export default (state) => Restore.create(state, actions)

/* globals fetch */

import EventEmitter from 'events'
import Restore from 'react-restore'

import link from '../../resources/link'
import * as actions from './actions'

export default (state, cb) => {
  // console.log('state...', state)
  const store = Restore.create(state, actions)
  store.events = new EventEmitter()
  link.on('action', (action, ...args) => {
    if (store[action]) store[action](...args)
  })

  return store
}
