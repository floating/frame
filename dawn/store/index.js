/* globals fetch */

import EventEmitter from 'events'
import Restore from 'react-restore'

import link from '../../resources/link'
import * as actions from './actions'

export default (state, cb) => {
  const store = Restore.create(state, actions)
  link.on('action', (action, ...args) => {
    if (store[action]) store[action](...args)
  })
  store.events = new EventEmitter()
  return store
}
