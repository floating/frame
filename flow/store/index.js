// import Restore from 'react-restore'
// import * as actions from './actions'
// export default (state) => Restore.create(state, actions)

/* globals fetch */

import EventEmitter from 'events'
import Restore from 'react-restore'

import link from '../../resources/link'
import * as actions from './actions'

export default (state, cb) => {
  const store = Restore.create(state, actions)
  store.events = new EventEmitter()

  // Feed for relaying state updates
  store.api.feed((state, actions, obscount) => {
    actions.forEach((action) => {
      action.updates.forEach((update) => {
        if (update.path.startsWith('main')) return
        link.send('tray:syncPath', update.path, update.value)
      })
    })
  })

  link.on('action', (action, ...args) => {
    if (store[action]) store[action](...args)
  })

  return store
}
