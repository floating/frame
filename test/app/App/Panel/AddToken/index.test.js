import React from 'react'
import Restore from 'react-restore'
import { render, waitFor } from '@testing-library/react'

import { setupComponent, advanceTimers } from '../../../../componentSetup'
import store from '../../../../../main/store'
import link from '../../../../../resources/link'
import AddTokenComponent from '../../../../../dash/App/Tokens/AddToken'

jest.mock('../../../../../main/store/persist')
jest.mock('../../../../../resources/link', () => ({
  invoke: jest.fn().mockResolvedValue({}),
  send: jest.fn()
}))

const AddToken = Restore.connect(AddTokenComponent, store)

beforeAll(() => {
  store.addNetwork({
    id: 1,
    type: 'ethereum',
    name: 'Mainnet',
    explorer: 'https://etherscan.io',
    symbol: 'ETH',
    on: true,
    connection: {
      primary: { connected: true }
    }
  })

  store.setPrimary('ethereum', 1, { connected: false })
  store.activateNetwork('ethereum', 1, true)

  store.addNetwork({
    id: 137,
    type: 'ethereum',
    name: 'Polygon',
    explorer: 'https://polygonscan.com',
    symbol: 'MATIC',
    on: true,
    connection: {
      primary: { connected: true }
    }
  })

  store.setPrimary('ethereum', 137, { connected: false })
  store.activateNetwork('ethereum', 137, true)
})

describe('selecting token chain', () => {
  it('should display the expected chain IDs', () => {
    const { getAllByRole } = render(<AddToken />)
  
    const tokenChainNames = getAllByRole('button').map((el) => el.textContent)
    expect(tokenChainNames).toEqual(['Mainnet', 'Polygon'])
  })
  
  it('should update add token navigation when a chain is selected', async () => {
    // 200 ms UI delay after clicking the button to select a chain
    const { user, getByRole } = setupComponent(<AddToken />, { advanceTimers: () => advanceTimers(200) })
  
    const polygonButton = getByRole('button', { name: 'Polygon' })
    await user.click(polygonButton)
  
    expect(link.send).toHaveBeenCalledWith('tray:action', 'navDash', {
      view: 'tokens',
      data: {
        notify: 'addToken',
        notifyData: {
          chainId: 137
        }
      }
    })
  })
})

describe('setting token address', () => {
  it('should prompt for a contract address if a chain has been selected', () => {
    const { getByLabelText } = render(<AddToken data={{ notifyData: { chainId: 137 }}} />)
  
    const contractAddressInput = getByLabelText(`Enter token's address`)
    expect(contractAddressInput.textContent).toBe('')
  })
  
  it('should update add token navigation when an address is entered', async () => {
    const { user, getByLabelText, getByRole } = setupComponent(<AddToken data={{ notifyData: { chainId: 137 }}} />)

    const contractAddressLabel = getByLabelText(`Enter token's address`)
    await user.type(contractAddressLabel, '0x3432b6a60d23ca0dfca7761b7ab56459d9c964d0')
    const setAddressButton = getByRole('button', { name: 'Set Address' })
    await user.click(setAddressButton)
  
    expect(link.send).toHaveBeenCalledWith('tray:action', 'navDash', {
      view: 'tokens',
      data: {
        notify: 'addToken',
        notifyData: {
          chainId: 137,
          address: '0x3432b6a60d23ca0dfca7761b7ab56459d9c964d0'
        }
      }
    })
  })

  it('should retrieve token metadata from a connected chain when an address is entered', async () => {
    store.setPrimary('ethereum', 137, { connected: true })

    const { user, getByLabelText, getByRole } = setupComponent(<AddToken data={{ notifyData: { chainId: 137 }}} />)

    const contractAddressLabel = getByLabelText(`Enter token's address`)
    await user.type(contractAddressLabel, '0x3432b6a60d23ca0dfca7761b7ab56459d9c964d0')
    const setAddressButton = getByRole('button', { name: 'Set Address' })
    await user.click(setAddressButton)
  
    expect(link.invoke).toHaveBeenCalledWith('tray:getTokenDetails', '0x3432b6a60d23ca0dfca7761b7ab56459d9c964d0', 137)
  })
})

describe('setting token details', () => {
  it('should prompt for default token details', () => {
    const { getByLabelText, getByRole } = render(
      <AddToken data={{ notifyData: { chainId: 137, address: '0x64aa3364F17a4D01c6f1751Fd97C2BD3D7e7f1D4' }}} />
    )
  
    const contractAddressInput = getByRole('heading')
    const tokenNameInput = getByLabelText('Token Name')
    const tokenSymbolInput = getByLabelText('Symbol')
    const tokenDecimalsInput = getByLabelText('Decimals')
  
    expect(contractAddressInput.textContent).toEqual('0x64aa3364D7e7f1D4')
    expect(tokenNameInput.value).toEqual('Token Name')
    expect(tokenSymbolInput.value).toEqual('SYMBOL')
    expect(tokenDecimalsInput.value).toEqual('?')
  })
  
  it('should update with loaded token metadata', async () => {
    store.setPrimary('ethereum', 137, { connected: true })
  
    const { user, getByLabelText, getByRole, rerender } = setupComponent(<AddToken data={{ notifyData: { chainId: 137 }}} />)
  
    link.invoke.mockResolvedValue({
      name: 'Frame Test on Polygon',
      symbol: 'mFRT',
      decimals: 18
    })
  
    link.send.mockImplementation(() => {
      rerender(<AddToken data={{ notifyData: { chainId: 137, address: '0x3432b6a60d23ca0dfca7761b7ab56459d9c964d0' }}} />)
    })
  
    const contractAddressLabel = getByLabelText(`Enter token's address`)
    await user.type(contractAddressLabel, '0x3432b6a60d23ca0dfca7761b7ab56459d9c964d0')
    const setAddressButton = getByRole('button', { name: 'Set Address' })
    await user.click(setAddressButton)
  
    const contractAddressInput = getByRole('heading')
    const tokenNameInput = getByLabelText('Token Name')
    const tokenSymbolInput = getByLabelText('Symbol')
    const tokenDecimalsInput = getByLabelText('Decimals')
  
    expect(contractAddressInput.textContent).toEqual('0x3432b6a6d9c964d0')
    await waitFor(() => expect(tokenNameInput.value).toEqual('Frame Test on Polygon'), { timeout: 200 })
    expect(tokenSymbolInput.value).toEqual('mFRT')
    expect(tokenDecimalsInput.value).toEqual('18')
  })
})
