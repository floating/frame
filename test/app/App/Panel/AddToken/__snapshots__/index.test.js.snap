// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should generate the expected HTML 1`] = `
<DocumentFragment>
  <div
    class="notifyBoxWrap cardShow"
  >
    <div
      class="notifyBoxSlide"
    >
      <div
        class="addTokenTitle"
      >
        Add New Token
      </div>
      <div
        class="addToken"
      >
        <div
          class="tokenRow"
        >
          <div
            class="tokenName"
          >
            <label
              class="tokenInputLabel"
            >
              Token Name
              <input
                class="tokenInput tokenInputAddress tokenInputDim"
                spellcheck="false"
                value="Token Name"
              />
            </label>
          </div>
        </div>
        <div
          class="tokenRow"
        >
          <div
            class="tokenSymbol"
          >
            <label
              class="tokenInputLabel"
            >
              Symbol
              <input
                class="tokenInput tokenInputAddress tokenInputDim"
                spellcheck="false"
                value="SYMBOL"
              />
            </label>
          </div>
          <div
            class="tokenDecimals"
          >
            <label
              class="tokenInputLabel"
            >
              Decimals
              <input
                class="tokenInput tokenInputAddress tokenInputDim"
                spellcheck="false"
                value="?"
              />
            </label>
          </div>
          <div
            class="tokenChainId"
          >
            <label
              class="tokenInputLabel"
            >
              Chain
              <div
                class="dropdownWrap"
              >
                <div
                  class="dropdown tokenInput"
                >
                  <div
                    class="dropdownItems"
                    role="listbox"
                    style="margin-top: 0px;"
                  >
                    <div
                      aria-selected="true"
                      class="dropdownItem"
                      role="option"
                      value="1"
                    >
                      Mainnet
                      <div
                        class="dropdownItemIndicator"
                      />
                    </div>
                    <div
                      aria-selected="false"
                      class="dropdownItem"
                      role="option"
                      value="137"
                    >
                      Polygon
                      <div
                        class="dropdownItemIndicator"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </label>
          </div>
        </div>
        <div
          class="tokenRow"
        >
          <div
            class="tokenAddress"
          >
            <label
              class="tokenInputLabel"
            >
              Contract Address
              <input
                class="tokenInput tokenInputAddress tokenInputDim"
                spellcheck="false"
                value="Contract Address"
              />
            </label>
          </div>
        </div>
        <div
          class="tokenRow"
        >
          <div
            class="tokenLogoUri"
          >
            <label
              class="tokenInputLabel"
            >
              Logo URI
              <input
                class="tokenInput tokenInputAddress tokenInputDim"
                spellcheck="false"
                value="Logo URI"
              />
            </label>
          </div>
        </div>
        <div
          class="tokenRow"
        >
          <div
            class="addTokenSubmit"
          >
            Fill in Token Details
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
